File: 2022/05/5.py
import re

[stacksinput, opsinput] = [s.splitlines() for s in open("input.txt").read().split("\n\n")]
crate_indices = [match.span()[0] for match in re.finditer(r"\d", stacksinput[-1])]

stacks = [list(filter(lambda s: s != " ", stack)) 
  for stack in zip(*[list(map(lambda i: row[i], crate_indices)) 
  for row in stacksinput[:-1]])]

ops = [[int(n), int(f) - 1, int(t) - 1]
  for n, f, t in [re.findall(r"\d+", op) for op in opsinput]]

def move(stacks, op, reverse=True):
  to_add = stacks[op[1]][0:op[0]]
  if reverse:
    to_add.reverse()
  stacks[op[2]] = to_add + stacks[op[2]]
  stacks[op[1]] = stacks[op[1]][op[0]:]
  return stacks

stacks2 = stacks.copy()

for op in ops:
  stacks = move(stacks, op)
  stacks2 = move(stacks2, op, reverse=False)

print("Part 1:", "".join([stack[0] for stack in stacks]))
print("Part 2:", "".join([stack[0] for stack in stacks2]))

File: 2022/05/5.clj
(ns aoc.2022.05.5
  (:require [aoc.common :refer [re-seq-indexed read-input zip]]
            [clojure.string :as str]))

(defn move [& {:keys [reverse]}]
  (fn [stacks [num from to]]
    (let [where (- (count (get stacks from)) num)
          to-add ((if reverse rseq identity) (subvec (get stacks from) where))]
      (-> stacks
          (update to #(reduce conj % to-add))
          (update from subvec 0 where)))))

(defn solve [stacks ops move]
  (->> ops
       (reduce move stacks)
       (map last)
       str/join))

(let [[stacks-input ops-input] (->> (read-input :split-with #"\n\n")
                                    (map str/split-lines))
      crate-indices (->> (last stacks-input)
                         (re-seq-indexed #"\d")
                         (map :start))
      stacks (->> stacks-input
                  drop-last
                  (map (fn [stack] (map #(.charAt stack %) crate-indices)))
                  (apply zip)
                  (map reverse)
                  (mapv (fn [l] (vec (take-while #(not= \space %) l)))))
      ops (->> ops-input
               (map #(re-seq #"\d+" %))
               (map #(map parse-long %))
               (map (fn [[a b c]] [a (dec b) (dec c)])))]

  (println "Part 1:" (solve stacks ops (move :reverse true)))
  (println "Part 2:" (solve stacks ops (move :reverse false))))